src/validation.cpp

```diff
@@ -4341,8 +4379,11 @@ void CChainState::CheckBlockIndex(const Consensus::Params& consensusParams)
     while (pindex != nullptr) {
         nNodes++;
         if (pindexFirstInvalid == nullptr && pindex->nStatus & BLOCK_FAILED_VALID) pindexFirstInvalid = pindex;
-        if (pindexFirstMissing == nullptr && !(pindex->nStatus & BLOCK_HAVE_DATA)) pindexFirstMissing = pindex;
-        if (pindexFirstNeverProcessed == nullptr && pindex->nTx == 0) pindexFirstNeverProcessed = pindex;
+        // Don't start being rigorous about missing block data if we're within a utxo snapshot
+        if (pindex->nHeight > snapshot_height) {
+            if (pindexFirstMissing == nullptr && !(pindex->nStatus & BLOCK_HAVE_DATA)) pindexFirstMissing = pindex;
+            if (pindexFirstNeverProcessed == nullptr && pindex->nTx == 0) pindexFirstNeverProcessed = pindex;
+        }
```

Comment:
Once the background validation has catched up with the snapshot_height, is `GetSnapshotHeight` going to return `std::nullopt` and those checks play out ?

Reply:
I'll make a note to ensure that `CompleteSnapshotValidation()` nulls out `m_snapshot_blockhash` so that what you're saying here is actually what happens. Otherwise we would just run the full suite of checks after restart when (what was previously) the snapshot chainstate gets loaded in as a "regular" IBD'd chainstate.

Reply:
In commit "validation: fix CheckBlockIndex for multiple chainstates" (d7804383748e85bda1be40cf0782a3ca2e457066)

re: https://github.com/bitcoin/bitcoin/pull/21526#discussion_r671490286

> I'll make a note to ensure that `CompleteSnapshotValidation()` (in #15606) nulls out `m_snapshot_blockhash` so that what you're saying here is actually what happens. Otherwise we would just run the full suite of checks after restart when (what was previously) the snapshot chainstate gets loaded in as a "regular" IBD'd chainstate.

This would be good to do but it does seem fragile that you have to clear a `CChainState` atttribute to make more checking happen. I think (again) more ideally in the future assumeutxo snapshot logic would be limited to `ChainStateManager`. Chain state objects could have a `CChainState::m_validation_starting_height` member that could be used here instead of `snapshot_height` and `ChainStateManager` could set, so this code doesn't have to know about `ChainStateManager` and ask it about snapshots.

Reply:
Totally agree @ryanofsky, and I've got some changes coming that will attempt to limit the degree to which the notion of snapshots leak out of ChainstateManager, but I want to point out two things:
- having an `m_validation_starting_height` member on CChainState would still have to be toggled after snapshot validation completes, so the fragility you rightfully point out would still be there, and
- whether we call it "snapshot_height" or "m_validation_starting_height" or "assumed_valid_before", we're going to have some notion of an assumed-valid region of the chain that parts of the system have to kind of know about to a certain extent. I think it's a legitimate argument that "assmed-valid" is maybe a more appropriate, general label than talking about snapshots, but fundamentally they're isomorphic at the moment AFAICT.

I think the only real way to avoid the fragility of the checks here is to, instead of relying on CChainState member variables, introduce a new block index status flag, `BLOCK_ASSUMED_VALID`, and have CheckBlockIndex et al gate on that. I'm going to look at potentially doing this in a future changeset because this involves to some extent revamping the assumeutxo changes more generally.

Reply:
> having an `m_validation_starting_height` member on CChainState would still have to be toggled after snapshot validation completes, so the fragility you rightfully point out would still be there

Right, but the point for me is to just locate the complexity in chainstatemanager instead of having it exposed to other parts of the code. I think in this case complexity is like a conserved quantity, you can't get rid of it but you can move it from one place to another. If complexity is encapsulated one place and not exposed other places that's the best outcome.

Also I guess I personally have a bias to wanting to name things after _how they are used_ or _what they mean_, not _where they came from_. Data can come from lots of places, and snapshots are a new source of data. But it adds mental overhead for pruning, indexing, and other validation code to know about the snapshot lifecycle or care where blocks came from when they don't need to, so something like "validated_starting_height" or "locally_validated_starting_height" seems easier to understand and reason about than "snapshot_height" in code that doesn't otherwise have to deal with snapshots.

`BLOCK_ASSUMED_VALID` could be good too. (I do have some skepticism that it conveys similar information, but not the same information, so might leave code making unclear assumptions like every block before the last assume valid block has the flag, and every block after doesn't have it. Also it would seem more natural to have a flag indicating what kind of validation was done, instead of what kind of validation was skipped, so maybe BLOCK_VALIDATED_FROM_GENESIS, BLOCK_VALIDATED_FROM_SNAPSHOT could be more descriptive.)

But everything after the first paragraph here is just differences of opinion about naming, and you should chose the names you like and ignore the feedback you don't like. I'm not trying to play architect here, just say what things I'm thinking and maybe generate some useful ideas.

Reply:
"Snapshot height" has now been completely removed as something this code is concerned about; I'll post a longer comment summarizing the changes, but we now rely on a block index flag called `BLOCK_ASSUMED_VALID` to change CheckBlockIndex() behavior as necessary. We now no longer have to worry about ChainstateManager variables to bring this behavior "back to normal," since that `BLOCK_ASSUMED_VALID` is removed from the corresponding pindex as blocks are validated.